src/seismometer/core/decorators.py

```diff
@@ -0,0 +1,232 @@
+
+        @wraps(func)
+        def wrapped_func(*args, **kwargs):
+            if not self.SEISMOMETER_CACHE_ENABLED:
+                # Skip any caching logic, pure pass through
+                return func(*args, **kwargs)
+
+            arg_hash = _hash_function_args(*args, **kwargs)
+            file_dir = self.cache_dir / func.__name__
+            filepath = file_dir / arg_hash
+            if filepath and filepath.is_file():
+                logger.debug(f"From cache: {filepath}")
+                return self.load_fn(filepath)
+            else:
+                logger.debug(f"Cache not found: {filepath}")
+                result = func(*args, **kwargs)
+                if self.return_type and not isinstance(result, self.return_type):
+                    raise ValueError(
+                        f"The function '{func.__name__}' did not return the expected"
+                        + f" type '{self.return_type.__name__}'"
+                    )
+                file_dir.mkdir(parents=True, exist_ok=True)
+                self.save_fn(result, filepath)
+                logger.debug(f"Saved to cahe: {filepath}")
+                return result
+
+        return wrapped_func
+
+    @property
+    def cache_dir(self):
+        """
+        Method to return the cache directory.
+        """
+        return self.SEISMOMETER_CACHE_DIR / self.cache_name
+
+    def clear(self):
+        """
+        Method to clear the cache directory.
+        """
+        shutil.rmtree(self.cache_dir, ignore_errors=True)
+        logger.debug(f"Cache cleared: {self.cache_name}")
+
+    @classmethod
+    def disable(cls):
+        """
+        Method to disable the cache.
+        """
+        cls.SEISMOMETER_CACHE_ENABLED = False
+        logger.debug("Cache disabled")
+
+    @classmethod
+    def enable(cls):
+        """
+        Method to disable the cache.
+        """
+        cls.SEISMOMETER_CACHE_ENABLED = True
+        logger.debug("Cache enabled")
+
+    @classmethod
+    def is_enabled(cls):
+        """
+        Method to check if the cache is enabled.
+        """
+        return cls.SEISMOMETER_CACHE_ENABLED
+
+    @classmethod
+    def clear_all(cls):
+        """
+        Method to clear all cache directories.
+        """
+        shutil.rmtree(cls.SEISMOMETER_CACHE_DIR, ignore_errors=True)
+        logger.debug(f"Cache cleared: {cls.SEISMOMETER_CACHE_DIR}")
+
+
+def _hash_function_args(*args, **kwargs):
+    """
+    Function to hash the arguments and keyword arguments of a function.
+
+    Parameters
+    ----------
+    *args : Any
+        arguments to hash
+    *kwargs : Any
+        keyword arguments to hash
+
+    Returns
+    -------
+    str
+        hash of the arguments and keyword arguments
+    """
+    hash_object = hashlib.md5()
+    for arg in args:
+        to_hash = _pandas_safe_hash(arg)
+        hash_object.update(str(to_hash).encode())
+    for key, arg in kwargs.items():
```

Comment:
Can you put this block undertest?

Also, I think you may need to sort kwargs.  It's good that different combinations of keys+values  are different (line1) but I'd expect different orders of the same keys+values to have the same hash (line2):
![image](https://github.com/epic-open-source/seismometer/assets/36572844/ddac0ccc-fbc7-482e-9f3c-8ab018f7c382)

Reply:
tests added for sorting, also included some tests around pd.Series (should be sorted by index) and pd.Index (unsorted, as if you are passing an index, the order might be important for caching. )